database.cc

```diff
@@ -1325,15 +1353,22 @@ future<mutation> database::do_apply_counter_update(column_family& cf, const froz
 
             tracing::trace(trace_state, "Reading counter values from the CF");
             return counter_write_query(m_schema, cf.as_mutation_source(), m.decorated_key(), slice, trace_state)
-                    .then([this, &cf, &m, m_schema, timeout, trace_state] (auto mopt) {
+                    .then([this, &cf, &m, &mopt_persist, m_schema, timeout, trace_state, return_previous_state] (auto mopt) {
                 // ...now, that we got existing state of all affected counter
                 // cells we can look for our shard in each of them, increment
                 // its clock and apply the delta.
                 transform_counter_updates_to_shards(m, mopt ? &*mopt : nullptr, cf.failed_counter_applies_to_memtable());
                 tracing::trace(trace_state, "Applying counter update");
+
+                // If current state has some deleted cells we should return the preimage to CDC.
+                // That's how CDC will know about ineffective writes, so it can omit them (don't log them).
+                if (return_previous_state || (mopt && m_schema->cdc_options().enabled() && has_deleted_counter_cell(*mopt))) {
+                    mopt_persist = std::move(mopt);
```

Comment:
what is `mopt_persist`?

Reply:
Yes, this is the state of counters before update (kinda 'preimage'), which used to be discarded, but now we need it, so it's moved to a variable with greater scope (more "persistent").

Reply:
Can it be given a better name or at least a comment near its definition?